numpy/core/src/multiarray/descriptor.c

```diff
@@ -1711,36 +1687,32 @@ _convert_from_bytes(PyObject *obj, PyArray_Descr **at)
                 }
             }
         }
-        return PyArray_DescrConverter(item, at);
+        return _convert_from_any(item);
     }
 
-    if (PyDataType_ISUNSIZED(*at) && (*at)->elsize != elsize) {
-        PyArray_DESCR_REPLACE(*at);
-        if (*at == NULL) {
+    if (PyDataType_ISUNSIZED(ret) && ret->elsize != elsize) {
+        PyArray_DESCR_REPLACE(ret);
+        if (ret == NULL) {
             return NPY_FAIL;
         }
-        (*at)->elsize = elsize;
+        ret->elsize = elsize;
     }
     if (endian != '=' && PyArray_ISNBO(endian)) {
         endian = '=';
     }
-    if (endian != '=' && (*at)->byteorder != '|'
-        && (*at)->byteorder != endian) {
-        PyArray_DESCR_REPLACE(*at);
-        if (*at == NULL) {
+    if (endian != '=' && ret->byteorder != '|' && ret->byteorder != endian) {
+        PyArray_DESCR_REPLACE(ret);
+        if (ret == NULL) {
             return NPY_FAIL;
         }
-        (*at)->byteorder = endian;
+        ret->byteorder = endian;
     }
-    return NPY_SUCCEED;
+    return ret;
 
 fail:
```

Comment:
I am wondering if we should rename `fail` to `datatype_not_understood`, `fail` could just as well mean that an error is already set for me.

Reply:
Perhaps in a future patch. Agreed that the name is confusing. This code is such a mess that its hard to keep things in manageable patches as is, I have maybe 5 different branches all cleaning up similar code locally.

Reply:
Sure, I am impressed you are managing to split up nicely at all...